File: Data Structures & Algorithms/valid-tree/submission-3.py
class Solution:
    def dfs(self, i: int, parent: int, vis: List[int], adj: List[List[int]]) -> bool:
        vis[i] = 1

        for j in adj[i]:
            if not vis[j]:
                if self.dfs(j, i, vis, adj):
                    return True
            elif j != parent:
                return True

        return False
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        adj = [[] for _ in range(n)]
        for x,y in edges:
            adj[x].append(y)
            adj[y].append(x)
        vis = [0] * n
        if(self.dfs(0,-1,vis,adj)):
            return False
        for i in range(n):
            if not vis[i]:
                return False
        return True

File: Data Structures & Algorithms/count-connected-components/submission-3.go
func countComponents(n int, edges [][]int) int {
    adj := make([][]int, n)

    cnt:=0
    for _,edge := range edges{
        a:=edge[0]
        b:=edge[1]
        adj[a]=append(adj[a],b)
        adj[b]=append(adj[b],a)
    }
    vis := make([]int, n)
    var dfs func(int)
    dfs = func(i int) {
        vis[i]=1
        for _,k := range adj[i]{
            if vis[k]!=1{
                dfs(k)
            }
        }
    }
    for i := 0; i < n; i++ {
        if vis[i]!=1{
            dfs(i)
            cnt+=1
        }
    }
    return cnt
}
